Make `enumerate` methods accept a callback that throws

thomasleese opened the pull request:
This updates the `enumerateAsArray` and `enumerateAsDict` methods to accept a row callback function that throws an error. If the function does throw an error it will be propogated to the caller of the `enumerate` method.

This is inspired by https://github.com/swiftcsv/SwiftCSV/pull/139 with the review comments addressed to hopefully get this functionality in to a release.

thomasleese committed:
Make `enumerate` methods accept a callback that throws

This updates the `enumerateAsArray` and `enumerateAsDict` methods to
accept a row callback function that throws an error. If the function
does throw an error it will be propogated to the caller of the
`enumerate` method.

This is inspired by https://github.com/swiftcsv/SwiftCSV/pull/139 with
the review comments addressed to hopefully get this functionality in to
a release.

Co-Authored-By: "Philip B." <145237+philipbel@users.noreply.github.com>

DivineDominion commented on SwiftCSV/ParsingState.swift:
```diff
@@ -20,12 +20,12 @@ struct ParsingState {
-    let finishRow: () -> Void
+    let finishRow: () throws -> Void
     let appendChar: (Character) -> Void
     let finishField: () -> Void
```
Genuine question, not a back-handed request:  What do you think about allowing all closures to throw? The public API makes sense this way. But it's a bit odd that eventually this only affects `finishRow` -- why can finishing a row fail, but not finishing a field?

DivineDominion reviewed (approved):
Thank you for the submission!

Just one comment now that this is fresh on your mind, a point to discuss, where I don't know a good answer myself at the moment.

thomasleese committed:
Allow other `ParsingState` closures to throw

This follows up on the previous commit which made the `finishRow`
closure of `ParsingState` allowed to throw an error, by updating the
other two closures of the private API to also throw an error.

Although this is not as neccessary for the private API, it keeps the
closures consistent, and there's no reason why one should throw and the
others should not.

DivineDominion commented:
Thank you! 🙌
